src/realm/object-store/sectioned_results.cpp

```diff
@@ -276,42 +255,57 @@ ResultsSection::ResultsSection(SectionedResults* parent, Mixed key)
 bool ResultsSection::is_valid() const
 {
     util::CheckedUniqueLock lock(m_parent->m_mutex);
+    if (!m_parent->is_valid())
+        return false;
+    // See if we need to recalculate the sections before
+    // searching for the key.
     m_parent->calculate_sections_if_required();
-    auto it = m_parent->m_sections.find(m_key);
-    return it != m_parent->m_sections.end() && m_parent->is_valid();
+    if (m_parent->m_sections.find(m_key) == m_parent->m_sections.end())
+        return false;
+
+    return true;
 }
 
 Mixed ResultsSection::operator[](size_t idx) const
 {
     util::CheckedUniqueLock lock(m_parent->m_mutex);
+    if (!m_parent->is_valid())
+        throw std::logic_error("The parent Results collection is not valid.");
```

Comment:
I think `Results::InvalidatedException` is probably the appropriate exception here.